src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action['type']) {
+      case 'addProperties':
+        for (const key in action.extraData) {
+          stateCopy[key] = action.extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
+        break;
+
+      default:
+      // console.log('Error, unknown type');
+    }
+
+    history.push({ ...stateCopy });
```

Comment:
Good job pushing a copy of the object to the `history` array at the end of each loop cycle, following the best practices to avoid mutations on the original state. [CHECKLIST ITEM 1]